torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -205,6 +196,64 @@ bool canPropagateShapeByRunningIt(Node* node) {
   return true;
 }
 
+bool PropagateTensorShapeOnNode(
+    Node * node, bool insert_expands, std::vector<TensorTypePtr> types);
+bool PropagateCompleteShapeOnNode(
+    Node * node, bool insert_expands, std::vector<CompleteTensorTypePtr> types);
+
+void PropagateCatShape(Node * cat_node) {
```

Comment:
Is this just moved, or are there substantive changes?

Reply:
`PropagateCatShape` is new, `PropagateCompleteShapeOnNode` is mostly a move with minor tweaks (like adding a return boolean, and removing propagation by running at the end), `PropagateTensorShapeOnNode` is now.

Reply:
Thanks. I'm not so sure why we need two top-level loops for propagation. I would have expected that this high level structure would be the same in both cases, and you would just either propagated detailed or less detailed information, based on whether or not you had more or less info.

Reply:
I'm not exactly sure what you mean. This is exactly what happens now. If all arguments have complete type information available, we will try to use the complete shape prop. If this fails, those arguments still have valid `TensorType`s, and so we'll try to use the less detailed function. Finally, if this one fails too, we try to resort to shape prop by running.